### ISSUE TITLE
Default and/or keyword args

### ISSUE THREAD

**Issue by simonbyrne** [JuliaParallel/MPI.jl#423]
To simplify the interface a bit, I was wondering whether we should provide default values and/or move some arguments to keywords:

- `comm`: default = `MPI.COMM_WORLD`
- `root`: default = 0
- `tag`: default = 0 for `MPI.Send`, `MPI_ANY_TAG` for `MPI.Recv`
-  `source` for `MPI.Recv`: default = `MPI_ANY_SOURCE`.

**Comment by jakebolewski** [comment]
This seems like a good change, although to be backwards compatible I would think you have to preserve the original function definitions which seems to be a complication.

**Comment by jkozdon** [comment]
I’ve always liked it when the MPI.jl interface matches the underlying c interface when possible. I realize it’s not a perfect match, but makes it easier when jumping in from the traditional hpc side of things.

**Comment by sloede** [comment]
I think we should be careful about differentiating between defaults that are "probably used by most people" and those that have some clear underlying rationale. For example, I would consider `comm = MPI.COMM_WORLD` and `root = 0` as two sensible defaults, since a) most codes really only use one communicator and b) root is special, and thus collective operations virtually always are controlled from root. OTOH, setting defaults for `tag` and `source` is something I'd oppose, since they really only work for `MPI.Send` and `MPI.Recv` but not, e.g., for `MPI.Isend`/`MPI.Irecv`. But even then, using defaults for the `tag` can lead to subtle errors that are really hard to track down (as I had to learn the hard way, and on more than one occasion :see_no_evil:).

To summarize, I think setting defaults for `root` and `comm` sounds reasonable, but I wouldn't do it for the other mentioned parameters. Re @jkozdon's objections, I think it would be OK if `root` and `comm` are given as positional arguments with default values, as this allows the hardened HPC professional to have the code look exactly like the C/C++ code while in many typical MPI scenarios the code looks a little less cluttered for those who wish to use it.

**Comment by jkozdon** [comment]
I was also taught that assuming `MPI_COMM_WORLD` as the communicator was just a bad coding practice (even if it was happens in 90% of my codes), as it can lead to subtle hard to track down bugs later.

**Comment by simonbyrne** [comment]
For comparison, mpi4py defines all the operations as methods on a `Comm` class (so there can't be a default communicator), and uses:
- [`Send` and `Isend` have default `tag = 0`](https://github.com/mpi4py/mpi4py/blob/3285cf9a06ed1916966a9e2a0af05defe1c3a40e/src/mpi4py/MPI/Comm.pyx#L263)
- [`Recv` and `Irecv` have default `source=ANY_SOURCE` and `tag = ANY_TAG`](https://github.com/mpi4py/mpi4py/blob/3285cf9a06ed1916966a9e2a0af05defe1c3a40e/src/mpi4py/MPI/Comm.pyx#L276)
- [collectives have default `root = 0`](https://github.com/mpi4py/mpi4py/blob/3285cf9a06ed1916966a9e2a0af05defe1c3a40e/src/mpi4py/MPI/Comm.pyx#L572)

I am sympathetic to the argument about  `comm`, but what would people think of making it either the first positional argument (which is [what PETSc uses](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MatCreateSeqAIJ.html#MatCreateSeqAIJ)) (e.g. `MPI.Reduce(MPI.COMM_WORLD, X, +)`) or as a non-optional keyword argument (e.g. `MPI.Reduce(X, +; comm=MPI.COMM_WORLD)`)?

**Comment by sloede** [comment]
Hm, on second thought, I second @jkozdon's argument about a default `comm`  being able to introduce subtle bugs. Also, I don't see the advantage of making the communicator the first or a keyword argument - what would be the benefit of that? It would have to offset the added confusion for experienced MPI users, as they cannot transfer their knowledge from other languages to Julia anymore.

**Comment by simonbyrne** [comment]
I guess the point is that the communicator is the common argument to all MPI communication operations, so it would make sense for it to be the first argument. We already mess up arguments anyway by combining all the buffer operations into a single argument and using return types; the keyword changes would have further deviations from the C signatures.

**Comment by sloede** [comment]
I wonder if this is a discussion that would be better suited for #395, or at least is closely related to it? To me, I feel it would be beneficial to "converts" from other languages to have a clean, very MPI-like "low-level" interface that sticks closely to the standard C API, while all other convenience improvements are rolled up in the high-level interface.

**Comment by simonbyrne** [comment]
Okay, so i think the main args i want to move to be keywords are those with integer args, as those are the easiest to get confused. In that case, what I think makes sense is:
- `Send`/`send` have `dest` as required keyword arg, and `tag=0` as optional keyword arg
- `Recv!`/`recv`/`Probe`/`Iprobe` have `source=any` and `tag=any`as optional keyword args
- `Sendrecv!` have `dest` as required, and `source`, `sendtag` and `recvtag` as optional
- collectives have `root=0` as optional keyword
- one-sided ops have have have `rank` as required, and `disp=0` as optional keywords
